paperscraper/utils.py

```diff
@@ -14,88 +14,82 @@
-        with `session = ThrottledClientSession(rate_limit=15)`
+    Rate-throttled client session.
 
+    Original source: https://stackoverflow.com/a/60357775
     """
 
-    MIN_SLEEP = 0.001
+    MAX_WAIT_FOR_CLOSE = 6.0  # sec
+    TIME_BASE = MAX_WAIT_FOR_CLOSE - 1  # sec
 
     def __init__(self, rate_limit: float | None = None, *args, **kwargs) -> None:
         """
         Initialize.
 
         Args:
-            rate_limit: Optional number of requests per second to throttle.
+            rate_limit: Optional number of requests per second to throttle. If left as
+                None, no throttling is applied.
             *args: Positional arguments to pass to aiohttp.ClientSession.__init__.
             **kwargs: Keyword arguments to pass to aiohttp.ClientSession.__init__.
         """
         super().__init__(*args, **kwargs)
-        self.rate_limit = rate_limit
+        self._rate_limit = rate_limit
         self._start_time = time.time()
         if rate_limit is not None:
-            if rate_limit <= 0:
-                raise ValueError("rate_limit must be positive")
-            self._queue: asyncio.Queue | None = asyncio.Queue(
-                maxsize=max(2, int(rate_limit))
-            )
-            self._fillerTask: asyncio.Task | None = asyncio.create_task(
-                self._filler(rate_limit)
-            )
+            queue_size = int(rate_limit * self.TIME_BASE)
+            if queue_size < 1:
+                raise ValueError(
+                    f"Rate limit {rate_limit} is too low for a responsive close, please"
+                    f" increase to at least {1 / self.TIME_BASE} requests/sec."
+                )
+            self._queue: asyncio.Queue | None = asyncio.Queue(maxsize=queue_size)
+            self._fillerTask: asyncio.Task | None = asyncio.create_task(self._filler())
         else:
             self._queue = None
             self._fillerTask = None
 
-    def _get_sleep(self) -> float | None:
-        if self.rate_limit is not None:
-            return max(1 / self.rate_limit, self.MIN_SLEEP)
-        return None
-
     async def close(self) -> None:
         """Close rate-limiter's "bucket filler" task."""
         if self._fillerTask is not None:
             self._fillerTask.cancel()
-            with contextlib.suppress(asyncio.TimeoutError):
-                await asyncio.wait_for(self._fillerTask, timeout=0.5)
+            await asyncio.wait_for(self._fillerTask, timeout=self.MAX_WAIT_FOR_CLOSE)
         await super().close()
 
-    async def _filler(self, rate_limit: float = 1) -> None:
+    async def _filler(self) -> None:
         """Filler task to fill the leaky bucket algo."""
-        if self._queue is None:
+        if self._rate_limit is None:
             return
+        queue = cast(asyncio.Queue, self._queue)
+        # Enough time (sec) to add at least 1 request to the queue
+        sleep_interval = max(1 / self._rate_limit, 1e-3)  # sec
```

Comment:
Edge case detail, but we're capped at 1k RPS with this.

Reply:
Fwiw, we're not capped at 1k RPS here, because what happens is the loop below this will add >1 requests (as opposed to ~1 request) per iteration. I have added this as a comment

This 1k RPS is here just to ensure the sleep is a meaningful time, so `asyncio` will tend to another "thread"

Reply:
great point!!!